### ISSUE TITLE
Metrics have bug for training.

### ISSUE THREAD

**Issue by somedadaism** [PyTorchLightning/metrics#637]
## 🐛 Bug

I am under the impression that metrics can not be shared between validation and training as `validation_epoch_end` resets them before `training_epoch_end` is run.  
As an end user this is highly counter intuitive to me.

#### Code sample

```
from pathlib import Path

import pytorch_lightning as pl
import torch
import torch.nn as nn
from torchmetrics import Accuracy
import pytorch_lightning.loggers as pl_loggers
import torch.nn as nn
from pytorch_lightning.callbacks import RichProgressBar
from torch.nn.modules.linear import Linear
from torchvision.models import resnet18
import pytorch_lightning as pl
from torch.utils.data import DataLoader, random_split
from torchvision import datasets, transforms

PROJECT_DIR = Path(".")


class MNISTDataModule(pl.LightningDataModule):
    def __init__(
        self,
        data_dir,
        batch_size=32,
        val_percentage=1 / 12,
        num_workers=20,
    ):
        super().__init__()
        self.data_dir = data_dir
        self.batch_size = batch_size
        self.val_percentage = val_percentage
        self.num_workers = num_workers

    def prepare_data(self):
        datasets.MNIST(root=self.data_dir, train=True, download=True)
        datasets.MNIST(root=self.data_dir, train=False, download=True)

    def setup(self, stage):
        transform = transforms.Compose(
            [
                transforms.ToTensor(),
                transforms.Normalize((0.1307,), (0.3081,)),
            ]
        )
        if stage in (None, "fit"):
            mnist_train = datasets.MNIST(root=self.data_dir, train=True, transform=transform)
            self.mnist_train, self.mnist_val = random_split(
                mnist_train,
                [
                    len(mnist_train) - int(len(mnist_train) * self.val_percentage),
                    int(len(mnist_train) * self.val_percentage),
                ],
            )
        if stage in (None, "test"):
            self.mnist_test = datasets.MNIST(root=self.data_dir, train=False, transform=transform)

    def train_dataloader(self):
        mnist_train = DataLoader(
            self.mnist_train, batch_size=self.batch_size, num_workers=self.num_workers
        )
        return mnist_train

    def val_dataloader(self):
        mnist_val = DataLoader(
            self.mnist_val, batch_size=self.batch_size, num_workers=self.num_workers
        )
        return mnist_val

    def test_dataloader(self):
        mnist_test = DataLoader(
            self.mnist_test, batch_size=self.batch_size, num_workers=self.num_workers
        )
        return mnist_test


class ClassificationModule(pl.LightningModule):
    def __init__(
        self, head_network: nn.Module, embedding_network: nn.Module, loss_func=None, metrics=None
    ):
        super().__init__()
        self.embedding_network = embedding_network
        self.full_network = nn.Sequential(self.embedding_network, head_network)
        self.loss_func = loss_func or nn.CrossEntropyLoss()
        self.metrics = metrics or nn.ModuleList([Accuracy()])

    def forward(self, x):
        y_pred = self.full_network(x)
        return y_pred

    def forward_embedding(self, x):
        y_pred = self.embedding_network(x)
        return y_pred

    def _update_metrics(self, logits, y, **log_kwargs):
        for metric in self.metrics:
            metric.update(logits, y)

    def _log_metrics(self, mode, **log_kwargs):
        for metric in self.metrics:
            self.log(f"\n{mode}/{metric.__class__.__name__.lower()}", metric, **log_kwargs)

    def training_step(self, batch, batch_idx):
        x, y = batch
        logits = self.forward(x)
        loss = self.loss_func(input=logits, target=y)
        self.log(
            f"\ntraining/loss",
            loss,
            on_step=False,
            on_epoch=True,
            prog_bar=False,
            logger=True,
        )
        self._update_metrics(logits, y)
        self._log_metrics(
            mode="training",
            on_step=True,
            on_epoch=True,
            prog_bar=True,
            logger=True,
        )
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        logits = self.forward(x)
        loss = self.loss_func(input=logits, target=y)
        self.log(
            f"\nvalidation/loss",
            loss,
            on_step=False,
            on_epoch=True,
            prog_bar=True,
            logger=True,
        )
        self._update_metrics(logits, y)
        self._log_metrics(
            mode="validation",
            on_step=True,
            on_epoch=True,
            prog_bar=True,
            logger=True,
        )
        return loss

    def test_step(self, batch, batch_idx):
        x, y = batch
        logits = self.forward(x)
        loss = self.loss_func(input=logits, target=y)
        self.log(
            f"\ntesting/loss",
            loss,
            on_step=False,
            on_epoch=True,
            prog_bar=True,
            logger=True,
        )
        self._update_metrics(logits, y)
        self._log_metrics(
            mode="testing",
            on_step=True,
            on_epoch=True,
            prog_bar=True,
            logger=True,
        )
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=3e-4)
        scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
            optimizer, T_max=self.trainer.max_epochs
        )
        return [optimizer], [scheduler]


if __name__ == "__main__":
    data_module = MNISTDataModule(
        PROJECT_DIR / "data", num_workers=20, batch_size=128, val_percentage=0.2
    )

    base_network = resnet18(pretrained=True)
    embedding_network = nn.Sequential(
        nn.Conv2d(1, 3, padding="same", kernel_size=3),
        base_network,
    )

    classification_module = ClassificationModule(
        head_network=Linear(1000, 10),
        embedding_network=embedding_network,
        loss_func=nn.CrossEntropyLoss(),
    )
    rich_bar = RichProgressBar()
    tb_logger = pl_loggers.TensorBoardLogger(
        save_dir=PROJECT_DIR / "logs",
        name="experiment",
        default_hp_metric=False,
    )
    trainer = pl.Trainer(
        gpus=1,
        callbacks=[rich_bar],
        max_epochs=20,
        logger=[tb_logger],
        check_val_every_n_epoch=1,
        num_sanity_val_steps=0,
    )
    trainer.fit(
        model=classification_module,
        datamodule=data_module,
    )
    trainer.test(datamodule=data_module)
```
### Expected behavior
```
Epoch 7    ━━━━━╸━━━━━━━━━━━━━━━━ 118/469 0:00:04 • 0:00:11 32.89it/s loss: 0.0144 v_num: 0 
                                                                      validation/loss: 0.041
                                                                      validation/accuracy_e…
                                                                      0.995                 
                                                                      training/accuracy_epo…
                                                                      0.0
```
Here `training/accuracy_epoch` is always 0. I would have expected it to be non-zero.

### Environment

- PyTorch Version `pytorch_lighting.__version__==1.5.2`. (Have not checked other versions.)

**Comment by SkafteNicki** [comment]
Hi @somedadaism,
As you state yourself we highly recommend that users use separate metrics for train, val and test to not mess up the state between them. Also stated as the first note on [this page](https://torchmetrics.readthedocs.io/en/latest/pages/overview.html) in the documentation. So instead of having
```python
self.metrics = metrics or nn.ModuleList([Accuracy()])
```
in the beginning, please exchange with
```python
self.train_metrics = deepcopy(metrics) or nn.ModuleList([Accuracy()])
self.val_metrics = deepcopy(metrics) or nn.ModuleList([Accuracy()])
self.test_metrics = deepcopy(metrics) or nn.ModuleList([Accuracy()])
```
then I would add a extra argument to your `_update_metrics` method:
```python
def _update_metrics(self, logits, y, stage, **log_kwargs):
    # stage should be one of ["train", "val", "test"]
    metrics = getattr(self, f"{stage}_metrics")
    for metric in metrics:
        metric.update(logits, y)
```
and a similar update to the `_log_metrics` method. This should fix the issues you are having.

**Comment by somedadaism** [comment]
Hi thanks for your help. Yes I have already fixed this accordingly.
Nonetheless the fact that `validation_end` comes before `training_end` will predictably result in user error.
You are correct to point to the documentation. What the documentation honestly should say in my opinion is that it is guaranteed with the current state of the framework to introduce error when trying to manage shared metrics yourself.

**Comment by SkafteNicki** [comment]
So I am going to agree with you that it is probably impossible to get it working correctly in lightning if you only is using a single instance of the metric. That is the reason we have the note in the docs.

However, `torchmetrics` is not only used with lightning but was also developed to be used with just native `torch`. Using just `torchmetrics+torch` it is absolutely possible to do train/val/test with a single metric instance because you have full control of the call order.

Based on this I would say that what we can do is to add an additional note in this page https://torchmetrics.readthedocs.io/en/latest/pages/lightning.html that is related specifically to logging in lightning that states that you should *always* use different metrics for the different stages.

**Comment by somedadaism** [comment]
Yes I think that is the best option.
I concur that using vanilla `torch` you can use the metrics in a shared manner.  
All in all this probably is more of an `pytorch_lighting` issue than a `torchmetrics` issue.
I will close my issue and remember to RTFM next time. 👍
